extensions/iceberg/src/main/java/io/deephaven/iceberg/layout/IcebergPartitionedLayout.java

```diff
@@ -0,0 +1,93 @@
+    private final Map<URI, IcebergTableLocationKey> cache;
+
+    private static IcebergTableLocationKey locationKey(
+            final URI fileUri,
+            final Map<String, Comparable<?>> partitions,
+            @NotNull final ParquetInstructions readInstructions) {
+        return new IcebergTableLocationKey(fileUri, 0, partitions, readInstructions);
+    }
+
+    /**
+     * @param tableSnapshot The {@link Snapshot} from which to discover data files.
+     * @param fileIO The file IO to use for reading manifest data files.
+     * @param partitionColumns The columns to use for partitioning.
+     * @param readInstructions The instructions for customizations while reading.
+     */
+    public IcebergPartitionedLayout(
+            @NotNull final Snapshot tableSnapshot,
+            @NotNull final FileIO fileIO,
+            @NotNull final String[] partitionColumns,
+            @NotNull final ParquetInstructions readInstructions) {
+        this.tableSnapshot = tableSnapshot;
+        this.fileIO = fileIO;
+        this.partitionColumns = partitionColumns;
+        this.readInstructions = readInstructions;
+
+        this.cache = new HashMap<>();
+    }
+
+    public String toString() {
+        return IcebergFlatLayout.class.getSimpleName() + '[' + tableSnapshot + ']';
+    }
+
+    @Override
+    public synchronized void findKeys(@NotNull final Consumer<IcebergTableLocationKey> locationKeyObserver) {
+        final Map<String, Comparable<?>> partitions = new LinkedHashMap<>();
+        try {
+            // Retrieve the manifest files from the snapshot
+            final List<ManifestFile> manifestFiles = tableSnapshot.allManifests(fileIO);
+            for (final ManifestFile manifestFile : manifestFiles) {
+                final ManifestReader<DataFile> reader = ManifestFiles.read(manifestFile, fileIO);
+                for (DataFile df : reader) {
+                    final URI fileUri = URI.create(df.path().toString());
+
+                    IcebergTableLocationKey locationKey = cache.get(fileUri);
+                    if (locationKey == null) {
+                        final PartitionData partitionData = (PartitionData) df.partition();
+                        for (int ii = 0; ii < partitionColumns.length; ++ii) {
+                            partitions.put(partitionColumns[ii], (Comparable<?>) partitionData.get(ii));
+                        }
+                        locationKey = locationKey(fileUri, partitions, readInstructions);
+                        if (!locationKey.verifyFileReader()) {
+                            continue;
+                        }
+                        cache.put(fileUri, locationKey);
+                    }
+                    locationKeyObserver.accept(locationKey);
+                }
+            }
+        } catch (final Exception e) {
+            throw new TableDataException("Error finding Iceberg locations under " + tableSnapshot, e);
+        }
+    }
+}
```

Comment:
This class and the FlatLayout are very clean but I guess we would need to eventually move to the URI code that would be added as part of the #5206 PR.